### ISSUE TITLE
Setting the sentinel in RcppArmadilloAs.h

### ISSUE THREAD

**Issue by binxiangni** [RcppCore/RcppArmadillo#134]
I am kind of confused of the line 105 - 107 in [RcppArmadilloAs.h](https://github.com/RcppCore/RcppArmadillo/blob/master/inst/include/RcppArmadilloAs.h). Why should we set the sentinel? What's the meaning of this piece of code? Could anyone give me a hint? 

`// Setting the sentinel`
`arma::access::rw(res.col_ptrs[(unsigned) dims[1] + 1]) = std::numeric_limits<arma::uword>::max();`

**Comment by eddelbuettel** [comment]
It's been a while ... but at some point I read up sparse matrices and their representation between the various formats (see eg [Wikipedia on Sparse Matrices](https://en.wikipedia.org/wiki/Sparse_matrix)) and as I recall this one is "recommended but not required".   The tuples of values defining a sparse matrix may be ended with sentinel, it is good practice to do so but may not be required.

The rcpp-devel list may be a better audience for this.  Some folks who actually work with sparse matrices (ie not really me) lurk there and could help.

What you could try is round up some of the sparse matrix packages (ie SparseM, spam, of course Matrix itself) and create a few matrices and see what you get back.   The do the same with Armadillo.

**Comment by binxiangni** [comment]
I'm working on the conversion of 9 subclasses of CsparseMatrix in package Matrix from R to C++. When I omit this piece of code and test the conversion, it works. That's why I am curious about the purpose of that piece. I'll try more matrices and see what will happen. 

Btw, I am done with a rough part of function "AS" for CsparseMatrix and now I can see the result in R with the help of sourceCpp. But I wanna see if the converted matrices are consistent with the representation in Armadillo. I know I should pass a R object using SEXP, but how can I write a matrix from R in SEXP in a C++ program and pass it to function "AS"? Hope my words make sense.

**Comment by eddelbuettel** [comment]
But Armadillo sparse matrices may also be used by, say, MLPACK.  And removing the sentinel may be an issue there.  

And sorry, can you rephrase that last question?

**Comment by binxiangni** [comment]
Sorry for confusion. I mean when I use AS to pass a matrix from R to C++(Armadillo), the matrix from R can be represented by a SEXP. But how can I write a SEXP to represent a R matrix in "int main()" part of a C++ program  to test the AS function I write? That is, I wanna have a R matrix in a C++ program.

**Comment by eddelbuettel** [comment]
For starters, let's call it `as<>()` to stick with established procedure, documentation and all.  There is no function `AS()`.

Second, you can't just drop R context into a standalone `main()` as you need R for R context.  If you want to embed R inside of a C++ program, [RInside](https://cran.r-project.org/package=RInside) is a good solution.   But that may make things too complicated for testing.

Lastly, I understand you want to test the C++ and R interface.  For that, I would work on the interface of C++ and R, with a package, using Rcpp.   You have RcppArmadillo.  Just fork it, and add your local code.

Apologies if I misunderstand what you are asking.  It's been a busy and hectic day so far.

**Comment by eddelbuettel** [comment]
I think we can close this as excellent progress has been made in #139, #140, #142, #144.
